// dmx-ts/dist/src/DMX.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DMX = void 0;
const events_1 = require("events");
const devices_1 = require("./devices");
const Events_1 = require("./models/Events");
class DMX extends events_1.EventEmitter {
    constructor(options) {
        var _a;
        super();
        this._universesByName = new Map();
        const devices = (_a = options === null || options === void 0 ? void 0 : options.devices) !== null && _a !== void 0 ? _a : {};
        this._devices = Object.assign({}, devices_1.PredefinedDevices, devices);
    }
    addUniverse(name, universe) {
        return __awaiter(this, void 0, void 0, function* () {
            yield universe.init();
            universe.on(Events_1.Events.update, (channels, extraData) => {
                this.emit(Events_1.Events.update, name, channels, extraData);
            });
            this._universesByName.set(name, universe);
            return universe;
        });
    }
    update(universeName, channels, extraData) {
        const universe = this._universesByName.get(universeName);
        if (universe === undefined) {
            throw new Error(`Universe ${universe} does not exist`);
        }
        universe.update(channels, extraData || {});
    }
    updateAll(universe, value) {
        var _a;
        (_a = this._universesByName.get(universe)) === null || _a === void 0 ? void 0 : _a.updateAll(value);
        this.emit(Events_1.Events.updateAll, universe, value);
    }
    universeToObject(universeKey) {
        const universe = this._universesByName.get(universeKey);
        const u = {};
        for (let i = 0; i < 512; i++) {
            u[i] = (universe === null || universe === void 0 ? void 0 : universe.get(i)) || 0;
        }
        return u;
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            for (const uni of this._universesByName.values()) {
                yield uni.close();
            }
            this.removeAllListeners();
        });
    }
}
exports.DMX = DMX;
