// dmx-ts/dist/src/drivers/abstract-serial-driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractSerialDriver = void 0;
const events_1 = require("events");
const serialport_1 = require("serialport");
class AbstractSerialDriver extends events_1.EventEmitter {
    constructor(serialPort, args) {
        super();
        this._intervalHandle = undefined;
        this._sendInterval = args.sendInterval;
        this._serialPortName = serialPort;
        this._serialPortOptions = args.serialPortOptions;
        this._universe = Buffer.alloc(513);
    }
    init() {
        return new Promise((resolve, reject) => {
            this._serialPort = new serialport_1.SerialPort(Object.assign(Object.assign({}, this._serialPortOptions), { path: this._serialPortName }), (err) => {
                if (!err) {
                    this.start();
                    resolve();
                }
                else {
                    reject(err);
                }
            });
        });
    }
    close() {
        this.stop();
        return new Promise((resolve, reject) => this._serialPort.close((err) => err ? reject(err) : resolve()));
    }
    get serialPort() {
        return this._serialPort;
    }
    get universeBuffer() {
        return this._universe;
    }
    start() {
        if (this._intervalHandle !== undefined) {
            throw new Error('Driver is already running.');
        }
        this._intervalHandle = setInterval(this.sendUniverse.bind(this), this._sendInterval);
    }
    stop() {
        if (this._intervalHandle !== undefined) {
            clearInterval(this._intervalHandle);
            this._intervalHandle = undefined;
        }
    }
    get(channel) {
        return this._universe[channel];
    }
    update(channels, extraData) {
        for (const c in channels) {
            this._universe[c] = channels[c];
        }
        this.emit('update', channels, extraData);
    }
    updateAll(value) {
        for (let i = 1; i <= 512; i++) {
            this._universe[i] = value;
        }
    }
}
exports.AbstractSerialDriver = AbstractSerialDriver;
