// dmx-ts/dist/src/drivers/bbdmx.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BBDMXDriver = void 0;
const events_1 = require("events");
const dgram_1 = __importDefault(require("dgram"));
const UNIVERSE_LEN = 512;
class BBDMXDriver extends events_1.EventEmitter {
    constructor(deviceId = '127.0.0.1', options = {}) {
        super();
        this.readyToWrite = true;
        this.interval = options.dmxSpeed ? (1000 / options.dmxSpeed) : 24;
        this.options = options;
        this.universe = Buffer.alloc(UNIVERSE_LEN + 1);
        this.host = deviceId;
        this.port = options.port || 9930;
        this.dev = dgram_1.default.createSocket('udp4');
        this.start();
    }
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            this.start();
        });
    }
    sendUniverse() {
        if (this.readyToWrite) {
            this.readyToWrite = false;
            let channel;
            let messageBuffer = Buffer.from(UNIVERSE_LEN.toString());
            for (let i = 1; i <= UNIVERSE_LEN; i++) {
                channel = Buffer.from(' ' + this.universe[i]);
                messageBuffer = Buffer.concat([messageBuffer, channel]);
            }
            this.dev.send(messageBuffer, 0, messageBuffer.length, this.port, this.host, () => {
                this.readyToWrite = true;
            });
        }
    }
    close() {
        this.stop();
        this.dev.close();
    }
    update(u, extraData) {
        for (const c in u) {
            this.universe[c] = u[c];
        }
        this.emit('update', u, extraData);
    }
    updateAll(v) {
        for (let i = 1; i <= UNIVERSE_LEN; i++) {
            this.universe[i] = v;
        }
    }
    get(c) {
        return this.universe[c];
    }
    start() {
        this.timeout = setInterval(this.sendUniverse.bind(this), this.interval);
    }
    stop() {
        if (this.timeout)
            clearInterval(this.timeout);
    }
}
exports.BBDMXDriver = BBDMXDriver;
