// dmx-ts/dist/src/drivers/null.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullDriver = void 0;
const events_1 = require("events");
class NullDriver extends events_1.EventEmitter {
    constructor(options = {}) {
        var _a;
        super();
        this._universe = Buffer.alloc(513, 0);
        this._interval = 1000 / ((_a = options === null || options === void 0 ? void 0 : options.dmxSpeed) !== null && _a !== void 0 ? _a : 1);
    }
    init() {
        return __awaiter(this, void 0, void 0, function* () {
            this.start();
        });
    }
    close() {
        this.stop();
    }
    update(u, extraData) {
        for (const c in u) {
            this._universe[c] = u[c];
        }
        this.logUniverse();
        this.emit('update', u, extraData);
    }
    updateAll(v) {
        for (let i = 1; i <= 512; i++) {
            this._universe[i] = v;
        }
    }
    get(c) {
        return this._universe[c];
    }
    logUniverse() {
        console.log(this._universe.slice(1));
    }
    start() {
        this._timeout = setInterval(() => {
            this.logUniverse();
        }, this._interval);
    }
    stop() {
        clearInterval(this._timeout);
    }
}
exports.NullDriver = NullDriver;
