// dmx-ts/dist/src/drivers/socketio.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketIODriver = void 0;
const socket_io_1 = require("socket.io");
const events_1 = require("events");
class SocketIODriver extends events_1.EventEmitter {
    constructor(options) {
        super();
        options = options || {};
        const port = options.port || 18909;
        const debug = options.debug || false;
        this.server = new socket_io_1.Server(port, { allowEIO3: true });
        this.server.on('connection', (socket) => {
            if (debug)
                console.info(`Client connected [id=${socket.id}]`);
            socket.on('disconnect', () => {
                if (debug)
                    console.info(`Client gone [id=${socket.id}]`);
            });
        });
        this.universe = Buffer.alloc(513, 0);
    }
    init() {
        return __awaiter(this, void 0, void 0, function* () {
        });
    }
    close() {
        this.server.close();
    }
    update(u, extraData) {
        for (const c in u) {
            this.universe[c] = u[c];
        }
        this.server.sockets.emit('update', [...this.universe]);
        this.emit('update', u, extraData);
    }
    updateAll(v) {
        for (let i = 1; i <= 512; i++) {
            this.universe[i] = v;
        }
        this.server.sockets.emit('update', [...this.universe]);
    }
    get(c) {
        return this.universe[c];
    }
}
exports.SocketIODriver = SocketIODriver;
